src/magic_folder/magic_folder.py

```diff
@@ -857,3 +874,97 @@ def add_file(self, path):
         return d
+
+@attr.s
+class RemoteSnapshotCreator(object):
+    _state_db = attr.ib()
+    _local_author = attr.ib()
+
+    @eliotutil.inline_callbacks
+    def _upload_local_snapshots(self, tahoe_client):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+
+        :param tahoe_client: a TahoeClient instance
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._state_db.get_all_localsnapshot_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
+        for name in localsnapshot_names:
+            action = UPLOADER_SERVICE_UPLOAD_LOCAL_SNAPSHOTS(relpath=name)
+            with action:
+                # deserialize into LocalSnapshot
+                snapshot = self._state_db.get_local_snapshot(name, self._local_author)
+
+                # now upload each item in the queue
+                try:
+                    remote_snapshot = yield write_snapshot_to_tahoe(
+                        snapshot,
+                        self._local_author,
+                        tahoe_client,
+                    )
+
+                    # At this point, remote snapshot creation successful for
+                    # the given relpath.
+                    # store the remote snapshot capability in the db.
+                    yield self._state_db.store_remotesnapshot(name, remote_snapshot)
+
+                    # Remove the local snapshot content from the stash area.
+                    snapshot.content_path.remove()
+
+                    # Remove the LocalSnapshot from the db.
+                    yield self._state_db.delete_localsnapshot(name)
+
+
+                except NoServersError:
+                    # Unable to reach Tahoe storage nodes because of
+                    # network errors or because the tahoe storage nodes
+                    # are offline. Retry?
+                    # XXX: Perhaps implement exponential backoff for retry?
+                    NO_NETWORK.log()
+                    continue
+                except Exception:
+                    write_traceback()
+
+@attr.s
+@implementer(service.IService)
+class UploaderService(service.Service):
+    """
+    A service that periodically polls the database for local snapshots
+    and commit them into the grid.
+    """
+    _tahoe_client = attr.ib()
+    _clock = attr.ib()
+    _polling_interval = attr.ib() # in GlobalConfigdatabase
```

Comment:
We might want to discuss how to handle configuration of these things "in general". That is, either we pass them the individual pieces of config they need upon creation (like here) or we pass them a config instance .. I don't see any code that actually creates one of these but one strategy would be a method like `UploaderService.from_config()` that takes a config but pulls out what the `UploaderService` actually needs.

I think I like that way, but it has follow-on implications for how we handle "the config has changed" -- with such a strategy, each service object would have to be re-created. Whereas if each thing was given its config, it would have to be told "your config (maybe) changed". I feel like the former strategy is straightforward and easy but I haven't thought too hard about it (and there might be edge-cases that aren't great).